hudi-flink-datasource/hudi-flink/src/main/java/org/apache/hudi/sink/bulk/RowDataKeyGen.java

```diff
@@ -167,7 +167,9 @@ public String getPartitionPath(RowData rowData) {
   }
 
   private static String getRecordKey(Object[] keyValues, String[] keyFields, boolean consistentLogicalTimestampEnabled) {
-    return KeyGenerator.constructRecordKey(Arrays.asList(keyFields), Arrays.stream(keyValues).map(value -> getTimestampValue(consistentLogicalTimestampEnabled, value)).collect(Collectors.toList()));
+    AtomicInteger index = new AtomicInteger(0);
```

Comment:
This is a hot method and `AtomicInteger` could introduce overhead.  Should this method and the caller be refactored so that `RowDataKeyGen.recordKeyProjection` can be used based on the record key field name instead of the array index?

Reply:
The method only operates over the full set of values currently so by requiring a function for this, there are tradeoffs when trying to unify the logic

Reply:
I'm converting this to a BiFunction that operates on the field name and index of field in the record key array to avoid the overhead here

Reply:
I see that `RowDataKeyGen.recordKeyProjection` is actually generated from `keyFields` and the logic is a bit convoluted in the current shape on master.  Regardless, the latest changes look good.